util/tracing/detect/otlp.go

```diff
@@ -93,3 +99,70 @@ func otlpMetricExporter() (sdkmetric.Exporter, error) {
 func deltaTemporality(_ sdkmetric.InstrumentKind) metricdata.Temporality {
 	return metricdata.DeltaTemporality
 }
+
+type otlpTracerFactory struct{}
+
+func (t otlpTracerFactory) AutoConfigure() bool {
+	return getEnv("", "OTEL_EXPORTER_OTLP_ENDPOINT", "OTEL_EXPORTER_OTLP_TRACES_ENDPOINT") != ""
+}
+
+func (t otlpTracerFactory) New() (sdktrace.SpanProcessor, error) {
+	proto := getEnv("grpc", "OTEL_EXPORTER_OTLP_TRACES_PROTOCOL", "OTEL_EXPORTER_OTLP_PROTOCOL")
+
+	var c otlptrace.Client
+	switch proto {
+	case "grpc":
+		c = otlptracegrpc.NewClient()
+	case "http/protobuf":
+		c = otlptracehttp.NewClient()
+	// case "http/json": // unsupported by library
+	default:
+		return nil, errors.Errorf("unsupported otlp protocol %v", proto)
+	}
+
+	sp, err := otlptrace.New(context.Background(), c)
+	if err != nil {
+		return nil, err
+	}
+	return sdktrace.NewBatchSpanProcessor(sp), nil
+}
+
+type otlpMetricFactory struct{}
+
+func (o otlpMetricFactory) AutoConfigure() bool {
+	return os.Getenv("OTEL_EXPORTER_OTLP_ENDPOINT") != "" ||
+		os.Getenv("OTEL_EXPORTER_OTLP_METRICS_ENDPOINT") != ""
+}
+
+func (o otlpMetricFactory) New() (sdkmetric.Reader, error) {
+	proto := getEnv("grpc", "OTEL_EXPORTER_OTLP_METRICS_PROTOCOL", "OTEL_EXPORTER_OTLP_PROTOCOL")
```

Comment:
I believe OTel changed the default from gRPC -> http/protobuf, so it _might_ make sense to prefer that here since metrics are a new addition?

(BuildKit tracing support was added at a time when gRPC _was_ the default)

Reply:
Yea looks like this is the case. We likely will need to be explicit with our own.

https://opentelemetry.io/docs/specs/otel/protocol/exporter/

I'm going to see if the `semconv` package has the defaults so we aren't using so many magic strings for this.

Reply:
There's no constant for the default in semconv (seems to be just for attribute names) and I don't see anything in the `otlptrace` package. I'm also a bit concerned this might qualify as a breaking change. I do think we should change it to keep the defaults correct according to the spec although we'll need to point it out in the release notes.